src/gs-update-monitor.c

```diff
@@ -150,6 +192,269 @@ no_updates_for_a_week (GsUpdateMonitor *monitor)
+					 NULL);
+	gs_plugin_loader_job_process_async (helper->monitor->plugin_loader,
+					    plugin_job, NULL,
+					    (GAsyncReadyCallback) app_update_finished_cb,
+					    helper);
+}
+
+static void
+download_now_cb (GObject *obj,
+                 GParamSpec *pspec,
+                 UpdateScheduleHelper *helper)
+{
+	GsApp *app = helper->app;
+	gboolean download_now = mwsc_schedule_entry_get_download_now (helper->entry);
+	AsAppState state = gs_app_get_state (helper->app);
+	const gchar *app_id = gs_app_get_unique_id (app);
+
+	g_debug ("Got download-now=%s for updating app %s",
+		 (download_now ? "TRUE" : "FALSE"), app_id);
+
+	if (download_now) {
+		/* verify again if the app needs to be updated, if not,
+		 * unschedule the update */
+		if (state == AS_APP_STATE_UPDATABLE_LIVE) {
+			update_app (helper);
+		} else {
+			g_debug ("Should update app %s but its state is %s! "
+				 "Unscheduling the update...", app_id,
+				 as_app_state_to_string (state));
+
+			/* unschedule the update */
+			g_hash_table_remove (helper->monitor->scheduled_updates,
+					     app_id);
+		}
+		return;
+	}
+
+	if (state == AS_APP_STATE_INSTALLING) {
+		/* if we cannot update at the moment, cancel any automatically
+		 * started update */
+		if (!app_is_auto_updating (helper->app))
+			return;
+
+		g_debug ("Cancelling automatic update of app %s, as "
+			 "download-now is FALSE", app_id);
+		g_cancellable_cancel (gs_app_get_cancellable (helper->app));
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+async_result_cb (GObject *source_object,
+                 GAsyncResult *result,
+                 GAsyncResult **out_result)
+{
+	*out_result = g_object_ref (result);
+}
+
+static gboolean
+setup_scheduler (GsUpdateMonitor *monitor,
+		 GCancellable *cancellable,
+		 GError **error)
+{
+	g_autoptr(GAsyncResult) new_result = NULL;
+	GMainContext *context = NULL;
+	MwscScheduler *scheduler = NULL;
+
+	if (monitor->scheduler != NULL)
+		return TRUE;
+
+	context = g_main_context_default ();
+
+	mwsc_scheduler_new_async (cancellable,
+				  (GAsyncReadyCallback) async_result_cb,
+				  &new_result);
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
+
+	scheduler = mwsc_scheduler_new_finish (new_result, error);
+
+	if (scheduler == NULL)
+		return FALSE;
+
+	g_signal_connect_swapped (scheduler, "invalidated",
+				  (GCallback) monitor_clear_scheduler,
+				  monitor);
+	monitor->scheduler = scheduler;
+
+	return TRUE;
+}
+
+static void
+scheduled_entry_invalidated_cb (MwscScheduleEntry *entry,
+				const GError *error,
+				UpdateScheduleHelper *helper)
+
+{
+	const gchar *app_id = gs_app_get_unique_id (helper->app);
+
+	g_debug ("Removing scheduled update of app %s", app_id);
+	g_hash_table_remove (helper->monitor->scheduled_updates, app_id);
+}
+
+static void
+schedule_update (GsUpdateMonitor *monitor,
+		 GsApp *app,
+		 GCancellable *cancellable)
```

Comment:
Nitpick: Alignment problems with the arguments here?

Reply:
Tabs... :disappointed: